### PR TITLE
Add modexp precompile

### LINKED ISSUE #4: Implement modexp

### PR DESCRIPTION
**Pull request by gballet**
Closes #4.

### CODE REVIEW

**Comment by gballet** [comment]
PR to fix the broken build: https://github.com/ewasm/ewasm-precompiles/pull/43

**Comment by axic** [comment]
@gballet do you want to merge this or not before it works on Rinkeby?

**Comment by gballet** [comment]
Yeah, I would prefer to finish a rinkeby sync, it's safer.

**Comment by axic** [comment]
@gballet are we close merging this?

**Comment by gballet** [comment]
frankly, no. I am at about 1/3rd of a Rinkeby sync, due to an issue with sha256, and there is at this point only one contract that uses modexp. I need more data to be able to merge this one. The ecrecover one is looking much better though, found a couple bugs that I have fixed and I expect it to be mergeable soon.

**Comment by axic** [comment]
Well, you could just run the test ethereum test suite with it.. :)

**Comment by gballet** [comment]
already done

**Comment by axic** [comment]
> already done that, and yet rinkeby doesn't sync

If rinkeby shows other edge cases not covered by the test suite then it would make sense creating test cases for them.

**Comment by gballet** [comment]
@axic we are good to go

File: expmod/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+
+    /* padded on 32 bytes */
+    let mut padded = vec![0u8; (32 - (data.len() % 32)) % 32];
+    padded.append(&mut data);
+    padded
+}
+
+#[cfg(not(test))]
+#[no_mangle]
+pub extern "C" fn main() {
```
Comment (axic) [inline expmod/src/lib.rs]: I think the main part of testing is about decoding the inputs, which is done in this function.

I'd prefer to split this into function which can be tested using unit tests.

File: expmod/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+    const BASE_LENGTH_OFFSET: usize = 0;
+    const BASE_LENGTH_LENGTH: usize = 32;
+    const EXP_LENGTH_OFFSET: usize = BASE_LENGTH_LENGTH;
+    const EXP_LENGTH_LENGTH: usize = 32;
+    const MOD_LENGTH_OFFSET: usize = EXP_LENGTH_OFFSET + EXP_LENGTH_LENGTH;
+    const MOD_LENGTH_LENGTH: usize = 32;
+    const BASE_OFFSET: usize = MOD_LENGTH_OFFSET + MOD_LENGTH_LENGTH;
+
+    let input_size = ewasm_api::calldata_size();
+    // Geth will consider the input size to be 0 if it's less than
```
Comment (axic) [inline expmod/src/lib.rs]: Is there a test in the official suite for this? I'm not sure I've seen this in the EIP, though I haven't looked at it for a while.

Comment (gballet) [reply expmod/src/lib.rs]: it's definitely not in the white paper, but it's in the codebase. Which EIP are you talking about?

Comment (gballet) [reply expmod/src/lib.rs]: most of it depends on the output of `calldata_copy` and I suspect the BigInt library to be very inefficient so I don't want to pass those BigInts back and make things worse. Although it hasn't been submitted yet, I have tests for these in the geth PR.

File: expmod/src/lib.rs (hunk 1 above)
Comment (gballet) [reply expmod/src/lib.rs]: most of it depends on the output of calldata_copy and I suspect the BigInt library to be very inefficient so I don't want to pass those BigInts back and make things worse. Although it hasn't been submitted yet, I have tests for these in the geth PR.

File: expmod/src/lib.rs (hunk 2 above)
Comment (axic) [reply expmod/src/lib.rs]: The one which introduced it: https://eips.ethereum.org/EIPS/eip-198

Comment (axic) [reply expmod/src/lib.rs]: Basically this code in geth could be thought as an optimisation. The first 96 bytes contains the data lengths, but calldata is expected to be infinitely zero padded. In case it is less than 96 bytes, no matter what length is stored in the first fields, the values will be zero anyway.

This assumes that `modexp(0, 0, 0) = 0`.

However, one part this could cause a problem is gas calculation.

File: Cargo.toml [hunk 3]
```diff
@@ -8,6 +8,7 @@ members = [
     "ecpairing",
     "ecrecover",
     "ed25519",
+    "expmod",
```
Comment (axic) [inline Cargo.toml]: While the YP calls this expmod, every single client implementation (and the tests, including [geth](https://github.com/ethereum/go-ethereum/blob/master/core/vm/contracts.go#L149)) call it modexp.

Can we name it modexp?

File: expmod/src/lib.rs (hunk 2 above)
Comment (axic) [reply expmod/src/lib.rs]: Here's the go piece: https://github.com/ethereum/go-ethereum/blob/master/core/vm/contracts.go#L173

Comment (gballet) [reply expmod/src/lib.rs]: I'm aware of that piece of code since this is where this comment comes from. The gas calculation is correct once this line is in, since this is one of the things that are needed for a correct sync of Rinkeby (meaning, I compared all txs on rinkeby one by one with the original, and got the same result). Therefore, could you explain what your point is?

File: Cargo.toml (hunk 3 above)
Comment (gballet) [reply Cargo.toml]: we can :smile:

File: expmod/src/lib.rs (hunk 2 above)
Comment (gballet) [reply expmod/src/lib.rs]: just one extra point: this isn't an optimization because if it was, it would return `[0, 0, ... 0]` (32 times) whereas with this it will return an empty array. I skimmed through the EIP and it seems to me that this case isn't explicitely covered.

File: Cargo.toml (hunk 3 above)
Comment (gballet) [reply Cargo.toml]: and we did

**Comment by axic** [comment]
@gballet are we close to merge this?

File: modexp/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+fn f(x: usize) -> u64 {
+    let cost = match x {
+        0...64 => x * x,
+        65...1024 => (x * x / 4) + 96 * x - 3072,
+        _ => (x * x / 16) + 480 * x - 199680,
+    };
+    cost as u64
+}
+
+const G_QUAD_DIVISOR: u64 = 20;
```
Comment (jakelang) [inline modexp/src/lib.rs]: Maybe move this to the top, right under the `use` statements

File: modexp/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+use std::cmp;
+use std::ops::{Rem, Shl};
+
+fn f(x: usize) -> u64 {
+    let cost = match x {
+        0...64 => x * x,
+        65...1024 => (x * x / 4) + 96 * x - 3072,
+        _ => (x * x / 16) + 480 * x - 199680,
+    };
+    cost as u64
```
Comment (jakelang) [inline modexp/src/lib.rs]: I think this expression can be shortened, although it is personal preference.
Something like 
```
match x {
...
} as u64
```

File: modexp/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+    use num::cast::ToPrimitive;
+    use std::cmp::min;
+
+    const BASE_LENGTH_OFFSET: usize = 0;
+    const BASE_LENGTH_LENGTH: usize = 32;
+    const EXP_LENGTH_OFFSET: usize = BASE_LENGTH_LENGTH;
+    const EXP_LENGTH_LENGTH: usize = 32;
+    const MOD_LENGTH_OFFSET: usize = EXP_LENGTH_OFFSET + EXP_LENGTH_LENGTH;
+    const MOD_LENGTH_LENGTH: usize = 32;
+    const BASE_OFFSET: usize = MOD_LENGTH_OFFSET + MOD_LENGTH_LENGTH;
```
Comment (jakelang) [inline modexp/src/lib.rs]: I think it would be better to define all the constants in one place, e.g. at the top of the file or inside the relevant funcs.
For readability.

File: expmod/src/lib.rs (hunk 1 above)
Comment (axic) [reply expmod/src/lib.rs]: This is a trick some other precompiles use ([for example](https://github.com/ewasm/ethereum-bn128.rs/blob/master/src/lib.rs)
```
let mut padded_input = input.chain(io::repeat(0));
```

**Comment by gballet** [comment]
@axic we're good to go
